### DISCUSSION TITLE
Blazor unified project design, interesting read!

### DISCUSSION THREAD

**Discussion by leigh-pointer** [oqtane/oqtane.framework#2972] [General]
https://github.com/dotnet/aspnetcore/issues/49079

**Comment by sbwalker** [comment]
@leigh-pointer thank you for posting. I have been following along and there are a lot of aspects to consider. One item which is important to note is that .NET 8 is expected to be backward compatible - which means existing Blazor apps  are expected to work on .NET 8 without significant modification. However, if you want to take advantage of new features in .NET such as server-side rendering (SSR) then changes will be required. There is a LOT to consider when it comes to Oqtane. 

One of the biggest challenges is routing - Oqtane uses a custom router rather than the default Blazor router. The new SSR functionality relies on the server populating a route table with all known routes and then passing it to the client-side router so it can make decisions on whether a route should be considered "interactive" or not. In a standard Blazor app the UI routes are all statically defined using ```@page``` attributes. In Oqtane the routes are dynamic - managed in the database, and can be modified at run-time. In addition, Oqtane routes are multi-tenant. So there are some major challenges to consider. 

Some of the other items mentioned in the thread you posted are interesting, as Oqtane already provides some of this functionality (obviously using its own approach). For example, Oqtane already supports both Blazor Server and Blazor WebAssembly in  the same application (users can specify different hosting models for different sites in their installation). Oqtane also already allows you to build modules which run on all hosting models (this is because modules are built using a client/server architecture approach)... and it accomplishes this with a single output format (ie. a single set of module DLLs will run on Server or WebAssembly). Some of the  ideas in that thread will make things a LOT more complicated, as they are talking about producing multiple output formats (which will be more difficult to  manage).

Considering that it is already July and .NET 8 is expected to ship by the end of this year, it seems like the window of time for Microsoft to deliberate on these design decisions is quickly dwindling.

**Comment by sbwalker** [comment]
I added a comment to the thread: https://github.com/dotnet/aspnetcore/issues/49079#issuecomment-1620271260

**Comment by sbwalker** [comment]
@leigh-pointer more information related to Blazor Unified in .NET 8... the initial design which enables SSR in .NET 8 is the IRouteStateProvider ([#47759](https://github.com/dotnet/aspnetcore/pull/47759)). The IRouteStateProvider basically passes the populated route table from the server to the client which enables the client to determine which routes should be ignored (ie.  processed only server-side). 

The problem with this design is that the route table in .NET is static. This might be acceptable for traditional Web API routes but for app UI routes it is extremely limiting (and is the reason why Oqtane does not use the default Blazor router - Oqtane uses a custom router with more dynamic capabilities instead). 

To provide some more context:

- with static routes you cannot add a new page to your site at run-time
- with static routes you cannot change a page name in your site at run-time
- with static routes you cannot change  a page path or reorganize pages in a site at run-time
- with static routes you cannot remove a page from your site at run-time

All of the actions above require a code change, application deployment, and application restart when using static routing.  

I am hopeful that Microsoft will reconsider their approach and provide support for dynamic routes in Blazor Unified in .NET 8, as I think SSR is a desirable feature for Oqtane.

**Reply by leigh-pointer** [reply]
@sbwalker I totally get where you're coming from with the static routes issue. It's a real pain, especially when it comes to app UI routes. The lack of dynamism can be a massive roadblock for applications like Oqtane, and having to make code changes and do full app deployments for even small route changes is a bit of a headache.

Here's to hoping that Microsoft takes these concerns seriously and revisits their approach. Dynamic routes would undoubtedly make SSR in Blazor Unified way more attractive and user-friendly. I'm crossing my fingers for Oqtane!

**Reply by sbwalker** [reply]
Microsoft appears to be doing some experimentation here: https://github.com/dotnet/aspnetcore/compare/main...javiercn/razor-components-dynamic-routing
